destreamer.ts

```diff
@@ -1,226 +1,223 @@
-            guid = url.split('/').pop();
-        }
-        catch (e)
-        {
-            console.error(`Could not split the video GUID from URL: ${e.message}`);
-            process.exit(25);
-        }
-        if (guid) {
-            videoGuids.push(guid);
-        }
-    }
-    
-    console.log(videoGuids);
-    return videoGuids;
-}
-
-
-async function downloadVideo(videoUrls: string[], outputDirectory: string, session: Session) {
-    console.log(videoUrls);
-    const videoGuids = extractVideoGuid(videoUrls);
-    
-    console.log('Fetching title and HLS URL...');
-    let metadata: Metadata[] = await getVideoMetadata(videoGuids, session);
-    await Promise.all(metadata.map(async video => {
-        video.title = sanitize(video.title);
-        console.log(colors.blue(`\nDownloading Video: ${video.title}\n`));
-        
-        // Very experimental inline thumbnail rendering
-        await drawThumbnail(video.posterImage, session.AccessToken);
-        
-        console.info('Spawning ffmpeg with access token and HLS URL...');
-
-        const outputPath = outputDirectory + path.sep + video.title + '.mp4';
-
-        ffmpeg()
-          .input(video.playbackUrl)
-          .inputOption([
-            '-headers', `Authorization:\ Bearer\ ${session.AccessToken}`
-          ])
-          .format('mp4')
-          .saveToFile(outputPath)
-          .on('codecData', data => {
-            console.log(`Input is ${data.video} with ${data.audio} audio.`);
-          })
-          .on('progress', progress => {
-            console.log(progress);
-          })
-          .on('error', err => {
-            console.log(`ffmpeg returned an error: ${err.message}`);
-          })
-          .on('end', () => {
-            console.log(`Download finished: ${outputPath}`);
-          });
-
-    }));
-}
-
-
-function sleep(ms: number) {
-    return new Promise(resolve => setTimeout(resolve, ms));
-}
-
-
-async function main() {
-    sanityChecks();
-    let session = tokenCache.Read();
-    if (session == null)
-    {
-        session = await DoInteractiveLogin(argv.username);
-    }
-
-    downloadVideo(argv.videoUrls as string[], argv.outputDirectory, session);
-}
-
-main();
\ No newline at end of file
+import { TokenCache } from './TokenCache';
+import { getVideoMetadata } from './Metadata';
+import { Metadata, Session } from './Types';
+import { drawThumbnail } from './Thumbnail';
+
+import { execSync } from 'child_process';
+import puppeteer from 'puppeteer';
+import colors from 'colors';
+import fs from 'fs';
+import path from 'path';
+import yargs from 'yargs';
+import sanitize from 'sanitize-filename';
+import ffmpeg from 'fluent-ffmpeg';
+
+
+/**
+ * exitCode 25 = cannot split videoID from videUrl
+ * exitCode 27 = no hlsUrl in the API response
+ * exitCode 29 = invalid response from API
+ * exitCode 88 = error extracting cookies
+ */
+
+let tokenCache = new TokenCache();
+const loginUrl = "https://web.microsoftstream.com/";
```

Comment:
We're relying on one implementation at a particular point in time. Zero guarantees that `web` subdomain will still exist tomorrow. Also private tenants never see that subdomain.
I propose we navigate to the first video URL we pick up from CLI and let Open ID Connect redirect us to whatever the identity provider might be (could be ADFS or other crazy idea).

I believe this is the only way to be robust about the logging in process with minimal assumptions.

Reply:
Ok going to change it this way